evm/vm/message.py

```diff
@@ -187,12 +193,29 @@ def __init__(self,
             validate_access_list(access_list)
         self.access_list = access_list
 
+    @property
+    def paygas_total_fee(self):
+        return self.paygas_gasprice * self.transaction_gas_limit
+
+    @property
+    def paygas_gasprice(self):
+        if self._paygas_gasprice is not None:
+            return self._paygas_gasprice
+        else:
+            return 0
+
+    @paygas_gasprice.setter
+    def paygas_gasprice(self, gas_price):
+        if self.paygas_gasprice == 0:
```

Comment:
It seems like there should be a logic branch here which throws an exception (not a `VMException` but something that will bubble) if the value is already set, because if I understand correctly, we shouldn't even try to set the value if it's already set.

Reply:
What about returning `True` of `False` in `paygas_gasprice.setter` to indicate if the value is successfully set or not.
```python
if self.paygas_gasprice is None:
    self._paygas_gasprice = gas_price
    return True
else:
    return False
```

and in `PAYGAS` opcode check against this returned value and push `1` or `0` to the stack to indicate the outcome of this `PAYGAS` operation?
```python
def paygas(computation):
    ...
    if computation.msg.depth == 0 and computation.msg.paygas_gasprice(gas_price):
        …
        computation.stack.push(1)
    else:
        computation.stack.push(0)
```

It's like fail of value transfer in a call. It seems better to me that it does not result in failure of transaction but I'm not that confident about this. Could it be the situation that gas were paid in top level call but later during some calls to other contract, the contract trigger `PAYGAS` again?

Reply:
I like pushing 1 on success and 0 on failure.

Reply:
So my suggestion was not focused on VM semantics, but rather our python API semantics.  Here's the situation I'm trying to avoid.

```python
# somewhere we set the gas price.
msg.paygas_gasprice = 10

# elsewhere in the code, we try to set it again.
msg.paygas_price = 100  # this will silently fail
```

What I would like is that the later case cannot silently fail, hence, suggesting that we raise an exception.  Then, if we ever have a location in the code which tries to set the gas price when it's already set which doesn't explicitely handle whatever exception we raise, then we'll get a hard failure in our code, rather than a silent failure that only manifests later at some point when the gas accounting is off.

Reply:
@NIC619 

Look here: https://github.com/ethereum/py-evm/blob/master/evm/computation.py#L316-L338

Anything inside of vm execution which throws any exception that is a subclass of `VMError` gets handled there, resulting in `computation.is_error` being `True`.

What I'm proposing is something that throws something like a `TypeError` or another custom exception that **is not** a `VMError` subclass.  This way, if we write code which sets `msg.paygas_gasprice` without implementing error handling **and** the `msg.paygas_gasprice` has already been set... an error will be thrown in such a way that it won't be caught in the linked code above.

Does that make sense?